File: test/src/commonTest/kotlin/Callbacks.kt

import natives.test.SimpleCallback
import natives.test.callbackReturn
import natives.test.simpleCallback
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class Callbacks {

    @Test
    fun simpleTest() = withLib {
        var a = 0
        simpleCallback {
            a = it
        }
        assertEquals(2, a)
    }

    @Test
    fun callbackEquality() = withLib {
        val callback: SimpleCallback = {  }
        val result = callbackReturn(callback)
        assertEquals(callback, result)
    }

    @Test
    fun callbackReturnString() = withLib {
        assertTrue(natives.test.callbackReturnString { "test" })
    }

    @Test
    fun callbackPingString() = withLib {
        assertTrue(natives.test.callbackPingString { it })
    }

    @Test
    fun callbackPingCallback() = withLib {
        val item = { _: Int -> }
        val response = natives.test.callbackPingCallback({ it }, item)
        assertEquals(item, response)
    }
}

File: test/natives/test/src/callbacks.c
#include <api.h>
#include <string.h>
#include <stdlib.h>

// Callback

void simpleCallback(SimpleCallback* callback) {
    INVOKE(callback, 2);
}

SimpleCallback* callbackReturn(SimpleCallback* callback) {
    return callback;
}

bool callbackReturnString(StringCallback* callback) {
    const char* text = INVOKE(callback);
    bool result = strcmp(text, "test") == 0;
    free((void*)text);
    return result;
}

bool callbackPingString(StringPingCallback* callback) {
    return strcmp(INVOKE(callback, "test"), "test") == 0;
}

SimpleCallback* callbackPingCallback(CallbackPingCallback* callback, SimpleCallback* item) {
    return INVOKE(callback, item);
}